go-controller/pkg/network-controller-manager/network_controller_manager.go

```diff
@@ -70,14 +68,127 @@ type NetworkControllerManager struct {
 	// used for leader election
 	identity string
 
-	// controller for all networks, key is netName of net-attach-def, value is *Controller
-	// this map is updated either at the very beginning of ovnkube-master when initializing the default controller
-	// or when net-attach-def is added/deleted. All these are serialized and no lock protection is needed
-	ovnControllers map[string]NetworkController
+	defaultNetworkController BaseNetworkController
+
+	// net-attach-def controller handle net-attach-def and create/delete network controllers
+	nadController *netAttachDefinitionController
+}
+
+func (cm *networkControllerManager) NewNetworkController(nInfo util.NetInfo,
+	netConfInfo util.NetConfInfo) (NetworkController, error) {
+	cnci := cm.newCommonNetworkControllerInfo()
+	topotype := netConfInfo.GetTopologyType()
+	if netConfInfo.GetTopologyType() == ovntypes.Layer3AttachDefTopoType {
+		return ovn.NewSecondaryLayer3NetworkController(cnci, nInfo, netConfInfo), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topotype)
+}
+
+func (cm *networkControllerManager) NewDummyNetworkController(topoType, netName string) (NetworkController, error) {
+	cnci := cm.newCommonNetworkControllerInfo()
+	netInfo := util.NewNetInfo(&ovncnitypes.NetConf{NetConf: types.NetConf{Name: netName}, IsSecondary: true})
+	switch topoType {
+	case ovntypes.Layer3AttachDefTopoType:
+		return ovn.NewSecondaryLayer3NetworkController(cnci, netInfo, &util.Layer3NetConfInfo{}), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topoType)
+}
```

Comment:
I'm a bit confused about these 2 factory funcs ... When should they be used ?...

Are there flows that require using `NewNetworkController` or `NewDummyNetworkController` ? Which ?...

Reply:
DummyNetworkController is for the case to cleanup the leftover logical entities for networks that was deleted during ovnkube-master (kind of like syncNetworkAtttachmentDefintion()). But because the net-attach-def is deleted, we are only able to create a dummy networkController of that topology type (only has networkName, topology type) to cleanup the leftover logical entities.

Reply:
not sure what would be a good name for it, we can discuss tomorrow

Reply:
but I think dummyNetworkController gives more hint of the incompleteness of the controller.